lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java

```diff
@@ -367,6 +407,73 @@ public Class getClass(Schema schema) {
     }
   }
 
+  static final Package JAVA_LANG_PKG = Package.getPackage("java.lang");
+  static final String NS_MAP_ARRAY_RECORD =   // record name prefix
+    "org.apache.avro.reflect.Pair";
+  static final String NS_MAP_KEY = "key";     // name of key field
+  static final int NS_MAP_KEY_INDEX = 0;      // index of key field
+  static final String NS_MAP_VALUE = "value"; // name of value field
+  static final int NS_MAP_VALUE_INDEX = 1;    // index of value field
+
+  /*
+   * Non-string map-keys need special handling and we convert it to an
+   * array of records as: [{"key":{...}, "value":{...}}]
+   */
+  Schema createNonStringMapSchema(Type keyType, Type valueType,
+                                  Map<String, Schema> names) {
+    Schema keySchema = createSchema(keyType, names);
+    Schema valueSchema = createSchema(valueType, names);
+    Schema.Field keyField = 
+      new Schema.Field(NS_MAP_KEY, keySchema, null, null);
+    Schema.Field valueField = 
+      new Schema.Field(NS_MAP_VALUE, valueSchema, null, null);
+    String name = getNameForNonStringMapRecord(keyType, valueType,
+      keySchema, valueSchema);
+    Schema elementSchema = Schema.createRecord(name, null, null, false);
+    elementSchema.setFields(Arrays.asList(keyField, valueField));
+    Schema arraySchema = Schema.createArray(elementSchema);
+    return arraySchema;
+  }
+
+  /*
+   * Gets a unique and consistent name per key-value pair. So if the same
+   * key-value are seen in another map, the same name is generated again.
+   */
+  private String getNameForNonStringMapRecord(Type keyType, Type valueType,
+                                  Schema keySchema, Schema valueSchema) {
+    if (keyType instanceof Class && valueType instanceof Class) {
+        Class keyClass = (Class)keyType;
+        Class valueClass = (Class)valueType;
+        if (keyClass.getPackage().equals(JAVA_LANG_PKG) &&
+            valueClass.getPackage().equals(JAVA_LANG_PKG)) {
+            return NS_MAP_ARRAY_RECORD +
+                keyClass.getSimpleName() + valueClass.getSimpleName();
+        }
+    }
+
+    String name = keySchema.getFullName() + valueSchema.getFullName();
+    long fingerprint = 0;
+    try {
+      fingerprint = SchemaNormalization.fingerprint64(name.getBytes("UTF-8"));
+    } catch (UnsupportedEncodingException e) {
+      String msg = "Unable to create fingerprint for ("
+                   + keyType + ", "  + valueType + ") pair";
+      throw new AvroRuntimeException(msg, e);
+    }
+    if (fingerprint < 0) fingerprint = -fingerprint;  // ignore sign
+    String fpString = Long.toString(fingerprint, 16); // hex
+    return NS_MAP_ARRAY_RECORD + fpString;
+  }
+
+  static boolean isMapWithNonStringKeysSchema(Schema s) {
```

Comment:
Nit: Other places use _nonStringMap_, like _isNonStringMap_ so I think this would be better as _isNonStringMapSchema_ to match _createNonStringMapSchema_.